# Mexican Broadsides Data Analysis

For the upcoming Mexican Broadsides collection, we so far have an Excel sheet with call numbers and some metadata. These data will need to be converted, cleaned, and otherwise analyzed in order to produce quality metadata for the DAMS. 

## First step: data conversion
Step one was taking the primary sheet in Excel and converting that to a csv file via `csvkit`. The command was:

```
in2csv --sheet 'Mexican Broadsides Batch 1' Mexican_broadsides_batch_1.xlsx > broadsides.csv
```

In [21]:
import pandas as pd
import os
import glob

In [4]:
df1 = pd.read_csv('broadsides.csv')

In [5]:
df1.head()

BIB                CALL    LOC  \
0  b35481328   BX1427 .A35 1680z  cpn     
1  b35484287    BX1427 .A35 1767  cpn     
2  b35484718   BX1427 .A35 1770z  cpn     
3  b35484676   BX1427 .A35 1780z  cpn     
4  b35581864  BX1430.G7 C38 1845  cpn     

                                              AUTHOR  \
0                  Cofradía de San Benito de Palerma   
1  Cofradia del Santissimo Sacramento y Nuestra S...   
2  Cofradía de San Dimas y Nuestra Señora de la C...   
3  Cofradia de San Dimas y Nuestra Señora de la C...   
4  Catholic Church. Diocese of Guadalajara (Mexic...   

                                               TITLE  \
0  Patente de la Cofradia y Hermandad de la Coron...   
1  Patente de la Cofradia del Santissimo Sacramen...   
2  Patente de la Cofradia y Ermandad de el glorio...   
3  Patente de hermano de la ilustre Cofradia del ...   
4  El Ilmo. Sr. Dr. D. Diego Aranda, dignisimo ob...   

                                           PUBLISHED  \
0                   [Mexico] : [La Cofradia], [168-]   
1                  [Mexico] : [La Cofradia], [1767?]   
2                   [Mexico] : [La Cofradia], [177-]   
3                   [Mexico] : [La Cofradia], [178-]   
4  [Guadalajara, Mexico] : Imprenta de Rodriguez,...   

                                       GENRE   STATUS Notes  
0  Broadsides -- Mexico -- 1680-1689. rbgenr  In Aeon   NaN  
1  Broadsides -- Mexico -- 1760-1769. rbgenr  In Aeon   NaN  
2  Broadsides -- Mexico -- 1770-1779. rbgenr  In Aeon   NaN  
3  Broadsides -- Mexico -- 1780-1789. rbgenr  In Aeon   NaN  
4  Broadsides -- Mexico -- 1840-1849. rbgenr  In Aeon   NaN

In [6]:
bib = df1['BIB']

In [7]:
bibs = []
for line in bib:
    bibs.append(line.strip())

In [8]:
print(bibs)
print('\n',len(bibs))

['b35481328', 'b35484287', 'b35484718', 'b35484676', 'b35581864', 'b23614614', 'b66771638', 'b35485140', 'b32169991', 'b35323954', 'b35411545', 'b35567028', 'b35575542', 'b35575943', 'b35382752', 'b35562560', 'b35571275', 'b35572620', 'b35571652', 'b35574987', 'b35571196', 'b35571378', 'b35517323', 'b35573314', 'b35581736', 'b35586436', 'b35608080', 'b48687273', 'b29429821', 'b35581827', 'b35355062', 'b35354963', 'b35354811', 'b35509247', 'b35562730', 'b35572930', 'b33814508', 'b23614791', 'b35573740', 'b35567004', 'b34096279', 'b92452498', 'b92452383', 'b26409124', 'b35847499', 'b35382636', 'b57981590', 'b35485486', 'b35474166', 'b3549444x', 'b3547449x', 'b35514048', 'b35561440', 'b33844483', 'b35514954', 'b35517529', 'b35370725', 'b35498833', 'b35498808', 'b35563370', 'b35568549', 'b35568409', 'b35513172', 'b35515971', 'b35459116', 'b35572784', 'b3551100x', 'b35517311', 'b35519952', 'b35476709', 'b35513366', 'b35519332', 'b33022161', 'b23614560', 'b23614699', 'b35353053', 'b35352942'

So we see we have 155 lines or objects listed in the spreadsheet. Let's now look at the actual file directories to compare. Dive Into Python, specifically its chapter [Working with Directories](http://www.diveintopython.net/file_handling/os_module.html) has a lot of useful tips we can use:

In [9]:
os.chdir('/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files')
folders = [name for name in os.listdir(".") if os.path.isdir(name)]

In [10]:
print(folders[0:30])
print('\n',len(folders), "total folders")

['b35571652', 'b35411545', 'b35576418', 'b92452966', 'b35515971', 'b35577605', 'b92453156', 'b92849179', 'b34079567', 'b35459098', 'b92645847', 'b35352942', 'b35353922', 'b92472801', 'b92452383', 'b35589711', 'b92758587', 'b35484676', 'b35407773', 'b35474166', 'b34088465', 'b92826428', 'b35353053', 'b92478396', 'b92675153', 'b3551274x', 'b23614560', 'b3560122x', 'b35517529', 'b92850510']

 223 total folders


Hmmmm. So comparing above, there were only 155 objects listed, but the directory has 223.  

Because we can have 'complex objects', that is, objects that have more than 1 file, we should explore what that structure looks like

In [11]:
files = []
for f in folders:
    file = os.listdir(f)
    files.append(file)
    continue 

In [12]:
print(files[0:15])
print('\n',len(files), "total objects")

[['Thumbs.db', 'b35571652_1.tif'], ['b35411545_1.tif', 'Thumbs.db'], ['Thumbs.db', 'b35576418_1.tif'], ['b92452966_1.tif', 'Thumbs.db'], ['b35515971_1.tif', 'Thumbs.db'], ['b35577605_1.tif', 'Thumbs.db'], ['b92453156_1.tif', 'Thumbs.db'], ['Thumbs.db', 'b92849179_1.tif'], ['b34079567_2.tif', 'b34079567_1.tif', 'Thumbs.db'], ['b35459098_1.tif', 'Thumbs.db'], ['b92645847_1.tif', 'Thumbs.db'], ['b35352942_1.tif', 'Thumbs.db'], ['b35353922_1.tif', 'b35353922_2.tif', 'Thumbs.db'], ['b92472801_1.tif', 'Thumbs.db', 'b92472801_2.tif'], ['b92452383_2.tif', 'b92452383_1.tif']]

 223 total objects


See how within each group, there are brackets? This means we have made a "list of lists", which means that each group is its own list within that larger list. This is key to knowing how to filter it later.

So, we see that within our 223 objects, we have some that have 2 or more .tif files, and thankfully, the objects with only 1 .tif still maintain a `_1` suffix.  

But, we also have the junk files 'Thumbs.db' that we will want to filter out of the list within a list. we can use the `.remove` method, which removes items from lists

In [13]:
for file in files:
    while 'Thumbs.db' in file: file.remove('Thumbs.db')
print(files[0:15])
print('\n',len(files), "total objects")

[['b35571652_1.tif'], ['b35411545_1.tif'], ['b35576418_1.tif'], ['b92452966_1.tif'], ['b35515971_1.tif'], ['b35577605_1.tif'], ['b92453156_1.tif'], ['b92849179_1.tif'], ['b34079567_2.tif', 'b34079567_1.tif'], ['b35459098_1.tif'], ['b92645847_1.tif'], ['b35352942_1.tif'], ['b35353922_1.tif', 'b35353922_2.tif'], ['b92472801_1.tif', 'b92472801_2.tif'], ['b92452383_2.tif', 'b92452383_1.tif']]

 223 total objects


So now we have a clean file directory list for our objects. At the very beginning, we made a list from the spreadsheet. That 'bib' ID in the spreadsheet would be helpful to group this directory listing as a key. This means we could make a python dictionary, with the bib ID as the key, and the files that exist as the values 

In [14]:
listing_dict = dict(zip(folders, files))

In [17]:
dict(list(listing_dict.items())[0:15])

{'b23614791': ['b23614791_1.tif'],
 'b34077121': ['b34077121_1.tif'],
 'b34096279': ['b34096279_2.tif', 'b34096279_1.tif'],
 'b35354288': ['b35354288_2.tif', 'b35354288_1.tif'],
 'b35476709': ['b35476709_1.tif'],
 'b35519952': ['b35519952_1.tif'],
 'b35571378': ['b35571378_1.tif'],
 'b35572620': ['b35572620_1.tif'],
 'b35588913': ['b35588913_1.tif', 'b35588913_2.tif'],
 'b92453582': ['b92453582_1.tif'],
 'b92478359': ['b92478359_2.tif', 'b92478359_1.tif'],
 'b92483458': ['b92483458_1.tif', 'b92483458_2.tif'],
 'b92592533': ['b92592533_1.tif'],
 'b92675323': ['b92675323_1.tif', 'b92675323_2.tif'],
 'b92847419': ['b92847419_1.tif']}

## Alternative methods
We used some `os` functions to get directory listings, but as we saw, one side effect was getting junk files we then had to filter out. Could we find a more precise way to get just .tif files? It looks like `glob` might be able to, and it should sound familiar from bash. Let's see what the documentation says:

In [25]:
help(glob)

Help on module glob:

NAME
    glob - Filename globbing utility.

MODULE REFERENCE
    https://docs.python.org/3.5/library/glob.html
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

FUNCTIONS
    escape(pathname)
        Escape all special characters.
    
    glob(pathname, *, recursive=False)
        Return a list of paths matching a pathname pattern.
        
        The pattern may contain simple shell-style wildcards a la
        fnmatch. However, unlike fnmatch, filenames starting with a
        dot are special cases that are not matched by '*' and '?'
        patterns.
        
        If recursive is true, the pattern '**' will match any files and
        zero or more directories and subdirectories.
    
 

In [28]:
glob_list = glob.glob('/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/*/*.tif')
print(glob_list[0:25])

['/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b35571652/b35571652_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b35411545/b35411545_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b35576418/b35576418_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b92452966/b92452966_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b35515971/b35515971_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b35577605/b35577605_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b92453156/b92453156_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b92849179/b92849179_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b34079567/b34079567_2.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b34079567/b34079567_1.tif', '/mnt/digital-staging/Mexican-Broadsides/batch1/Working_Files/b35459098/b35459098_1.tif', '/mnt/dig